samples/net/sockets/echo_service/src/main.c

```diff
@@ -227,73 +250,89 @@ static int setup_udp_socket(struct sockaddr_in6 *addr)
 
 	if (bind(sock, (struct sockaddr *)addr, sizeof(*addr)) < 0) {
 		LOG_ERR("bind: %d", -errno);
+		close(sock);
 		return -errno;
 	}
 
 	return sock;
 }
 
-int main(void)
+int start_echo_service(void)
 {
 	int tcp_sock, udp_sock, ret;
-	char addr_str[INET6_ADDRSTRLEN];
 	struct sockaddr_in6 addr = {
 		.sin6_family = AF_INET6,
 		.sin6_addr = IN6ADDR_ANY_INIT,
 		.sin6_port = htons(MY_PORT),
 	};
-	static int counter;
 
 	client_list_init();
 
 	tcp_sock = setup_tcp_socket(&addr);
 	if (tcp_sock < 0) {
+		LOG_ERR("Failed to setup tcp listening socket");
 		return tcp_sock;
 	}
 
 	udp_sock = setup_udp_socket(&addr);
 	if (udp_sock < 0) {
-		return udp_sock;
+		LOG_ERR("Failed to setup udp socket");
+		ret = udp_sock;
+		goto cleanup_tcp;
 	}
 
-	sockfd_udp[0].fd = udp_sock;
-	sockfd_udp[0].events = POLLIN;
+	/* Register TCP listening socket to service handler */
+	struct pollfd tcp_server = { .fd = tcp_sock, .events = POLLIN };
+	ret = net_socket_service_register(&service_accept, &tcp_server, 1, NULL);
+	if (ret < 0) {
+		LOG_ERR("Cannot register socket service handler (%d)", ret);
+		goto cleanup_sockets;
+	}
 
 	/* Register UDP socket to service handler */
-	ret = net_socket_service_register(&service_udp, sockfd_udp,
-					  ARRAY_SIZE(sockfd_udp), NULL);
+	struct pollfd sockfd_udp = { .fd = udp_sock, .events = POLLIN };
+	ret = net_socket_service_register(&service_udp, &sockfd_udp, 1, NULL);
 	if (ret < 0) {
 		LOG_ERR("Cannot register socket service handler (%d)", ret);
+		goto cleanup_sockets;
 	}
 
+	k_mutex_lock(&lock, K_FOREVER);
+	tcp_socket = tcp_sock;
+	udp_socket = udp_sock;
+	k_mutex_unlock(&lock);
+
 	LOG_INF("Single-threaded TCP/UDP echo server waits "
 		"for a connection on port %d", MY_PORT);
 
-	while (1) {
-		struct sockaddr_in6 client_addr;
-		socklen_t client_addr_len = sizeof(client_addr);
-		int client;
-
-		client = accept(tcp_sock, (struct sockaddr *)&client_addr,
-				&client_addr_len);
-		if (client < 0) {
-			LOG_ERR("accept: %d", -errno);
-			continue;
-		}
-
-		inet_ntop(client_addr.sin6_family, &client_addr.sin6_addr,
-			  addr_str, sizeof(addr_str));
-		LOG_INF("Connection #%d from %s (%d)", counter++, addr_str, client);
+	return 0;
 
-		client_list_try_add(client);
+cleanup_sockets:
+	close(udp_sock);
+cleanup_tcp:
+	close(tcp_sock);
+	return -1;
 
-	}
+}
 
+int stop_echo_service(void)
```

Comment:
This doesn't seem to be used anywhere, perhaps we could use it to restart the services in case of more serious errors (like accept)? Perhaps it'd make sense to keep `main()` for that purpose after all (the resources are allocated in the system anyway, as main shares the thread with late initialization AFAIR).